### PR TITLE
Breaking changes

### LINKED ISSUE #23: Consider going back to base16 hash encoding
IPFS has a plan to migrate to using base32 as default encoding, see:
- https://github.com/ipfs/specs/issues/247
- https://github.com/ipfs/go-ipfs/issues/4143

Despite longer representation these hashes can be used in subdomains which are case-insensitive.

We should also reconsider our use of `base56btc` due to it being widely considered deprecated.

### LINKED ISSUE #42: Reconsider refs terminology
Borrowed git terminology is getting a bit confusing in combination with streams:
- is `head` a beginning of a stream or it's end?
- if latter, why `tail` command shows events closes to the `head`?
- how do we name an interval of blocks and in which order `[startBlock, endBlock]` are defined?

We could use better terms.

### LINKED ISSUE #46: Make `MetadataBlock::sequenceNumber` a `u64`

### LINKED ISSUE #57: Constrict `ReadStep::Csv`
We implemented CSV reader configuration to mimic [Spark reader options](https://spark.apache.org/docs/latest/sql-data-sources-csv.html). A lot of those options are not present in other libraries (e.g. Arrow) or can be replaced by a preprocess step.

We should consider dropping some of these options to make implementing the spec easier.

### LINKED ISSUE #59: Unify naming in `SetTransform`
`SetTransform` originally specified inputs by `name`, but later was extended to support multi-tenant references via `datasetRef`. We should consider removing `name`, `id` and leaving only `datasetRef` as it can support all three. In a normalized form we'd want `datasetRef` always be an ID and `alias` be the name used to refer to it withing transform's queries.

### LINKED ISSUE #60: [EPIC] Breaking changes
This is a tracking list of all braking changes we are making to the metadata in near future:

- [x] #46
- [x] #23
- [x] #59 
- [x] #61 
- [x] #57 
- [x] #42
- [x] #64
- [x] #69
- [x] #47 
- [x] #70

### LINKED ISSUE #61: Change `ExecuteQueryInput::datasetName` to `alias`
The `datasetName` field is restricted to dataset names, but this prevents it from carrying multi-tenant references like `<account>/<dataset>`. 

This should be done together with #59

### LINKED ISSUE #64: Make source names optional
When we decided to support multiple sources in `AddData::SourceState` we made this field required, and to represent polling sources used a `odf/polling` constant. Now that we are adding push sources - we don't want to force people giving them some random names (when you only have one source), so we want to make them optional in all places.

### LINKED ISSUE #69: Consider using half-open intervals for blocks and offsets

### LINKED ISSUE #70: Standardize enum representation in YAML/JSON encoding
Currently the spec is vague about how enum variants must be differentiated between when serialized in YAML/JSON form.

Kamu is using `kind: camelCaseVariantName`.

The proposal is, however, to use `kind: PascalCaseVariantName` so that the variant name was identical to how it appears in ODF schemas.

### PR DESCRIPTION
**Pull request by sergiimk**
A pack of breaking changes that closes #60 

- closes #46
  - Changed `MetadataBlock::sequenceNumber` type from `int32` to `uint64`
  - Similarly changed `size` fields from `int64` to `uint64`
  - Similarly change `offset` data type from `int64` to `uint64` - let's use types to the fullest
- closes #59
  - In `TransformInput` removed `id` and `name` properties in favor of `datasetRef` and (optional) `alias`
- closes #61 
  - In `ExecuteQueryInput`
    - rename `datasetName` to `queryAlias` and change format to allow any `string`
    - add `datasetAlias` (for logging purposes only)
    - rename `dataInterval` to `offsetInterval`
  - In `ExecuteQueryRequest`
    - rename `datasetName` to `datasetAlias` and change format to `dataset-alias` to support multi-tenant names (for logging purposes only)
  - rename `offset` to `nextOffset`
- closes #23
  - Recommend `base16` encoding for block hashes (see RFC-012)
- related to #6
  - Fix `did:odf` to encode a complete public key and not its hash (same as `did:key` method)

- closes #57
  - In `ReadStep::Csv` remove properties `enforceSchema`, `comment`, `emptyValue`, `nanValue`, `positiveInf`, `negativeInf`, `multiLine`, `ignoreLeadingWhiteSpace`, `ignoreTrailingWhiteSpace`
  - Removed deprecated `ReadStep::JsonLines` enum variant
- closes #70
  - Standardized enum representation in YAML/JSON encoding as `kind: PascalCaseSchemaName` (see RFC-013)
- closes #42
  - Made a choice of not renaming the `head` reference
- closes #64
  - Renamed `SourceState::source` to `sourceName` (and decided to keep it as required)
- closes #69
  - Propagation of last offset and last processed block through `AddData`, `ExecuteQuery`, and `ExecuteQueryInput` to minimize scanning
    - Rename `InputSlice` to `ExecuteQueryInput`, while renaming existing `ExecuteQueryInput` in engine ops to `ExecuteQueryRequestInput` to avoid conflict
    - Unify naming: instead of `input / output`, `last / next` pairs use `prev / new` consistently
    - Remove `BlockInterval` fragment
  - In `ExecuteQueryResponse::Success`
    - rename `dataInterval` to `offsetInterval`
  - In `InputSlice` rename `dataInterval` to `offsetInterval`
  - In `DataSlice` rename `interval` to `offsetInterval`

Other:
- Use `type: string, format: flatbuffers` in `SetDataSchema`, as in YAML for the schema is `base64`-encoded
- Clarified rules of propagating source states, checkpoints, and watermarks across events
- Unified to use `camelCase` for property names with "ID" in it (e.g. `datasetId` instead of `datasetID`)
- `DatasetSnapshot.name` can now be a multi-tenant alias
- Renames `ExecuteQuery` metadata event and the corresponding engine operation to `ExecuteTransform`
- Introduces `RawQuery` engine operation to be used primarily for batch preprocess step of the ingest (but possibly other uses too).

### CODE REVIEW

File: rfcs/003-content-addressability.md [hunk 1]
```diff
@@ -46,18 +46,17 @@ We will laso expand the use of `multihash` format proposed in [RFC-002](./002-lo
...
 Dataset creation procedure will involve:
 - Generating a new cryptographic key pair (defaulting to `ed25519` algorithm)
-- Computing a hash of the public key (defaulting to `sha3-256`)
-- Encoding it into `CIDv1` format (using appropriate codec code like `ed25519-pub`)
-- Storing this CID in the first Metadata Block's `seed` field.
+- Prefixing it with and appropriate `multicodec` identifier (like `ed25519-pub`)
+- Storing this data in the first Metadata Block's `seed` field.
 
-When representing dataset ID as a string the DID format `did:odf:<cid>` will be used, where CID uses mulitibase encoding.
+When representing dataset ID as a string the DID format `did:odf:<multibase>` will be used, where the binary data will use mulitibase format and `base58-btc` encoding just like in `did:key` method.
```
Comment (zaychenko-sergei) [inline rfcs/003-content-addressability.md:59]: mul**i**tibase format

File: rfcs/013-yaml-enum-representation.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+**Authors**:
+- [Sergii Mikhtoniuk](mailto:sergii.mikhtoniuk@kamu.dev), [Kamu](https://kamu.dev)
+
+**Compatibility**:
+- [ ] Backwards-compatible
+- [ ] Forwards-compatible
+
+## Summary
+Prpoposes to use `kind: PascalCaseVariantName` convention to standardize how enum variants are differentiated in YAML / JSON serialization format.
```
Comment (zaychenko-sergei) [inline rfcs/013-yaml-enum-representation.md:16]: Prpopo, hm :-)

File: schemas/metadata-events/ExecuteQuery.json [hunk 3]
```diff
@@ -5,33 +5,38 @@
...
       "type": "array",
       "items": {
-        "$ref": "/schemas/InputSlice"
+        "$ref": "/schemas/ExecuteQueryInput"
       },
-      "description": "Defines inputs used in this transaction. Slices corresponding to every input must be present."
+      "description": "Defines inputs used in this transaction. Slices corresponding to every input dataset must be present."
     },
-    "inputCheckpoint": {
+    "pevCheckpoint": {
```
Comment (zaychenko-sergei) [inline schemas/metadata-events/ExecuteQuery.json:18]: prev

**Review by zaychenko-sergei — CHANGES_REQUESTED** [review]

**Comment by zaychenko-sergei** [comment]
What's the motivation to rename `head` to `tail` ?

File: rfcs/013-yaml-enum-representation.md (hunk 2 above)
Comment (sergiimk) [reply rfcs/013-yaml-enum-representation.md:16]: I really need to install some spell checker :)

**Pull request merged**